### ISSUE TITLE
Is timeout handling possible without NCCL_BLOCKING_WAIT?

### ISSUE THREAD

**Issue by vadimkantorov** [NVIDIA/nccl#395]
https://pytorch.org/docs/stable/distributed.html#torch.distributed.init_process_group:
 ```
timeout (timedelta, optional) – Timeout for operations executed against the process group. Default value equals 30 minutes. This is applicable for the gloo backend. For nccl, this is applicable only if the environment variable NCCL_BLOCKING_WAIT is set to 1.
```

We found that always using NCCL_BLOCKING_WAIT leads to 20% slower training time. But timeout option is crucial for avoiding endless hanging and graceful exit of all processes when one of processes died for whatever reason (one frequent is OOM).

**Comment by phalexo** [comment]
Have you figured out where to change the timeout for it to work? Whatever I do, the training script dies around 30 minutes in.

**Comment by gkroiz** [comment]
This issue is old, but I'm curious @vadimkantorov if you found any alternative solutions for timeout handling.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.

**Comment by vadimkantorov** [comment]
> if you found any alternative solutions for timeout handling.

Unfortunately, not :(

@stephenmsachs are there any other ways for handling timeouts? this issues hasn't seen updates in over 6 months partly because nvidia has not provided any advise / response :(

**Comment by stephenmsachs** [comment]
Thanks for your feedback. IIUC unsetting `TORCH_NCCL_BLOCKING_WAIT` will unblock the current process and the [ncclCommWatchdogThread](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1016) will still kill the process if a timeout is reached.

**Comment by phalexo** [comment]
I figured out what the problem is in most of the cases. When a process pool
is used to prep data, often a fork is used. The problem is that when a fork
creates a new process, sometimes semaphores/mutex are created locked up.
Then other workers wait, and wait for the deadlocked worker and eventually
they time out.

Have to use fork server.

**Comment by vadimkantorov** [comment]
> will still kill the process if a timeout is reached.

@stephenmsachs In PyTorch docs is now written: `timeout (timedelta, optional) – Timeout for operations executed against the process group. Default value is 10 minutes for NCCL and 30 minutes for other backends. This is the duration after which collectives will be aborted asynchronously and the process will crash. This is done since CUDA execution is async and it is no longer safe to continue executing user code since failed async NCCL operations might result in subsequent CUDA operations running on corrupted data. When TORCH_NCCL_BLOCKING_WAIT is set, the process will block and wait for this timeout.`

It is still not clear if `timeout` will be respected if `TORCH_NCCL_BLOCKING_WAIT` is not set. My original question is whether it's possible to still have timeout respected without passing to `TORCH_NCCL_BLOCKING_WAIT=1` and its slowdowns. (the typical usecase 5 years ago was that one process crashes because of OOM, and other processes would wait indefinitely)

Do I understand you correctly that in recent PyTorch, the `timeout` is respected even without `TORCH_NCCL_BLOCKING_WAIT=1`?

**Comment by stephenmsachs** [comment]
> Do I understand you correctly that in recent PyTorch, the timeout is respected even without TORCH_NCCL_BLOCKING_WAIT=1?

Yes, that is how I understand the watchdog. And from experience I can tell you that it kills the processes when the timeout is reached.
